### PR TITLE
Fix the ProductSync's publish/unpublish build actions 

### PR DESCRIPTION
**Pull request by ahmetoz**
#### Summary
Check the task for the bug, #530 
#### Todo

- Tests
    - [X] Unit 
    - [X] Integration
- [-] Documentation
- [x] Add Release Notes entry.

#### Hints for Review

Check the fix first with following the cases in `BuildPublishOrUnpublishUpdateActionTest`
Some tests are adjusted, as some conditions before (state 13,14,15,16) was missing/bug.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 1]
```diff
@@ -508,25 +508,180 @@ private static boolean isSameForAllActionNew(
...
     @Nonnull
-    public static Optional<UpdateAction<Product>> buildPublishUpdateAction(@Nonnull final Product oldProduct,
-                                                                           @Nonnull final ProductDraft newProduct) {
+    public static Optional<UpdateAction<Product>> buildPublishOrUnpublishUpdateAction(
+        @Nonnull final Product oldProduct,
+        @Nonnull final ProductDraft newProduct,
+        final boolean hasNewUpdateActions) {
+
         final Boolean isNewProductPublished = toBoolean(newProduct.isPublish());
         final Boolean isOldProductPublished = toBoolean(oldProduct.getMasterData().isPublished());
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java:676]:
```suggestion
        final boolean isOldProductPublished = toBoolean(oldProduct.getMasterData().isPublished());
```

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 2]
```diff
@@ -508,25 +508,180 @@ private static boolean isSameForAllActionNew(
...
+
         if (Boolean.TRUE.equals(isNewProductPublished)) {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java:678]:
```suggestion
        if (isNewProductPublished) {
```

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 3]
```diff
@@ -508,25 +508,180 @@ private static boolean isSameForAllActionNew(
...
+            if (Boolean.TRUE.equals(isOldProductPublished)
+                && (hasNewUpdateActions || oldProduct.getMasterData().hasStagedChanges())) {
+
+                // covers the state 14, state 15 and state 16.
+                return Optional.of(Publish.of());
+            }
             return buildUpdateAction(isOldProductPublished, isNewProductPublished, Publish::of);
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java:685]:
```suggestion
            return buildUpdateAction(isOldProductPublished, true, Publish::of);
```

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 4]
```diff
@@ -508,25 +508,180 @@ private static boolean isSameForAllActionNew(
...
         }
         return buildUpdateAction(isOldProductPublished, isNewProductPublished, Unpublish::of);
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java:687]:
```suggestion
        return buildUpdateAction(isOldProductPublished, false, Unpublish::of);
```

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/products/ProductSyncIT.java [hunk 5]
```diff
@@ -409,7 +445,8 @@ void sync_withProductTypeReference_ShouldUpdateProducts() {
 
         assertThat(updateActions).containsExactlyInAnyOrder(
             SetAttributeInAllVariants.of(targetProductRefAttr.getName(), targetProductRefAttr.getValue(), true),
-            SetAttributeInAllVariants.of(targetProductSetRefAttr.getName(), targetProductSetRefAttr.getValue(), true)
+            SetAttributeInAllVariants.of(targetProductSetRefAttr.getName(), targetProductSetRefAttr.getValue(), true),
+            Publish.of()
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/products/ProductSyncIT.java:449]: What happened here in this test that we got additionally published action?

Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/products/ProductSyncIT.java:449]: Some tests are adjusted after the fix, this case was missing before. I did not change the current prepared products.

In this case, when the target product is published a product is updated, the product will `hasStagedChanges=true`, stage and current will be different and sync needs to send a publish to make sure changes are synced between staged and current projections.

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/products/templates/beforeupdatecallback/KeepOtherVariantsSyncIT.java [hunk 6]
```diff
@@ -129,6 +129,6 @@ void sync_withRemovedVariants_shouldNotRemoveVariants() {
 
         assertThat(productOptional).isNotEmpty();
         final Product fetchedProduct = productOptional.get();
-        assertThat(fetchedProduct.getMasterData().getCurrent().getVariants()).hasSize(1);
+        assertThat(fetchedProduct.getMasterData().getCurrent().getVariants()).hasSize(2);
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/products/templates/beforeupdatecallback/KeepOtherVariantsSyncIT.java:132]: It was wrong before. Similar case : https://github.com/commercetools/commercetools-sync-java/pull/533#discussion_r465620706

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java (hunk 1 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/533/commits/ffe106572cbca9358a19d1a5b446a3acd14d7d35

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java (hunk 2 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/533/commits/ffe106572cbca9358a19d1a5b446a3acd14d7d35

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java (hunk 3 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/533/commits/ffe106572cbca9358a19d1a5b446a3acd14d7d35

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java (hunk 4 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/533/commits/ffe106572cbca9358a19d1a5b446a3acd14d7d35

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 7]
```diff
@@ -508,25 +508,180 @@ private static boolean isSameForAllActionNew(
...
-     * Compares the 'published' field of a {@link ProductDraft} and a {@link Product} and accordingly returns
-     * a {@link Publish} or {@link Unpublish} update action as a result in an {@link Optional}. If the new product's
-     * 'published' field is null, then the default false value is assumed.
-     *
-     * <p>If both the {@link Product} and the {@link ProductDraft} have the same 'published' flag value, then no update
-     * action is needed and hence an empty {@link Optional} is returned.
-     *
-     * <p>NOTE: Comparison is done against the staged projection of the old product.
+     * Compares the 'published' field of a {@link ProductDraft} and a {@link Product} with the new update actions
+     * and hasStagedChanges of the old product accordingly returns a {@link Publish} or {@link Unpublish} update action
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/533/commits/d691a292cbc242fab96fd8a0bac55f5f3ea15b27

**Pull request merged**